src/object.cpp

```diff
@@ -1166,4 +1166,26 @@ StringObject *Object::to_str(Env *env) {
         result->klass()->inspect_str());
 }
 
+int Object::to_cmp_int(Env *env, Value lhs, Value rhs) {
+    if (is_nil()) {
+        env->raise_not_comparable_error(lhs, rhs);
+    }
+
+    if (is_integer()) {
```

Comment:
This conditional is unnecessary

Reply:
Actually, this method is confusing. It accepts `lhs` and `rhs` but doesn't use them??

Reply:
Yeah it looks odd! It accepts them as in Ruby the error message generated by values which can't be compared contains their class names. So lhs and rhs are forwarded to the error raise method. It's basically the same implementation as the one in Ruby. I'm not sure how strictly natalie follows the message formatting for errors of Ruby but I see these "not comparable" exceptions dotted around the code base and thought we could make a helper to generate them (as in Ruby)

Reply:
Hmm, I see. Still a bit of an odd method. It's just not intuitive how to use it.

I wonder if it would make more sense to remove `lhs` and use `this`.

Or, alternatively, move it off of Object, and make a regular function that takes `lhs` and `rhs`.

Reply:
Ok thanks for the suggestion @seven1m. It prob makes sense for the method to simply live as a standalone function.

If on Object, its usage would be odd if `lhs` was `this` as the value we are converting is not the `lhs` of the comparison but rather the return value of the thing used in the comparison, eg in the case of `Infinity`s, the `this` here will be the return value of `infinite?` that we are converting (while the `lhs` will be the Infinite itself) or in sort the return value of the block (ie in`.sort { |a, b| c }`,  `c` is the one being converted, while `a` and `b` are the `lhs` and `rhs`)

One option could be to see if that `nil` guard be removed thus negating the need for `lhs` and `rhs`... [Ruby has it](https://github.com/ruby/ruby/blob/5ccb625fbbd1e774636a9fdbe0bf1c3d38e085d5/bignum.c#L2955) but maybe it could be avoided for `nil` guards at the call sites. 

I'll have a look at its usages in Ruby and see what I can do!

Btw for context, here is a case where that `nil` guard comes into play:

```ruby
irb(main):002:0> [1, 2, 3, 4].sort { |a, b| nil}
Traceback (most recent call last):
	20: from /Users/stephen/.asdf/installs/ruby/2.7.2/bin/irb:23:in `<main>'
	19: from /Users/stephen/.asdf/installs/ruby/2.7.2/bin/irb:23:in `load'
	18: from /Users/stephen/.asdf/installs/ruby/2.7.2/lib/ruby/gems/2.7.0/gems/irb-1.4.2/exe/irb:11:in `<top (required)>'
	 1: from (irb):2:in `<main>'
(irb):2:in `sort': comparison of Integer with 2 failed (ArgumentError)
```

The return value from `sort` should be converted into the range [-1, 0 ,1]. Here the value `nil` is being returned from the block, so when passed to `rb_cmpint` (https://github.com/ruby/ruby/blob/5ccb625fbbd1e774636a9fdbe0bf1c3d38e085d5/array.c#L3492) causes the guard to be triggered and raises our error (the `lhs` and `rhs` in this case is just 2 of the integers of the array being compared).

In `natalie` the return value conversion to the necessary range is duplicated and handled inline in various places, eg in `sort` https://github.com/natalie-lang/natalie/blob/f85770c630afb19102b9fd1e595c6381b9702e21/src/array_object.cpp#L1028 (also as the TODO there indicates, the error isnt created how Ruby does it, as it does some logic on the `lhs` before raising as I tried to replicated in `Env::raise_not_comparable_error`)

Btw interestingly for `sort`, the result is different to Ruby here:

```
irb(main):011:0> val = Object.new; def val.<(other); false; end; def val.>(other); false; end; [1, 2, 3, 4].sort { |a, b| val }
=> [1, 2, 3, 4]
```

compared to:
```
TLW:natalie stephen$ ./bin/natalie -e " val = Object.new; def val.<(other); false; end; def val.>(other); false; end; puts [1, 2, 3, 4].sort { |a, b| val }"
[4, 1, 2, 3]
```

Ill see if I can find the root cause of this too!

Sorry for the long discussion!